src/Toml/Parser/TOML.hs

```diff
@@ -68,69 +68,58 @@ inlineTableP = between
     (text "{") (text "}")
     (tomlFromInline <$> hasKeyP `sepEndBy` text ",")
 
--- | Parser for a table.
-tableP :: Parser (Key, TOML)
-tableP = do
-    key  <- tableNameP
-    toml <- subTableContent key
-    pure (key, toml)
-
--- | Parser for an array of tables.
-tableArrayP :: Parser (Key, NonEmpty TOML)
-tableArrayP = do
-    key       <- tableArrayNameP
-    localToml <- subTableContent key
-    more      <- optional $ sameKeyP key tableArrayP
-    case more of
-        Nothing         -> pure (key, localToml :| [])
-        Just (_, tomls) -> pure (key, localToml <| tomls)
+
+-- | Parser for an array of tables under a certain key.
+tableArrayP :: Key -> Parser (NonEmpty TOML)
+tableArrayP key =
+    tableP (Just key) `NC.sepBy1` sameKeyP key tableArrayNameP
 
 -- | Parser for a '.toml' file
 tomlP :: Parser TOML
-tomlP = do
-    sc
-    (val, inline)  <- distributeEithers <$> many hasKeyP
-    (table, array) <- fmap distributeEithers
-        $ many
-        $ eitherPairP (try tableP) tableArrayP
-
-    pure TOML
-        { tomlPairs       = HashMap.fromList val
-        , tomlTables      = fromList $ inline ++ table
-        , tomlTableArrays = HashMap.fromList array
-        }
+tomlP = sc *> tableP Nothing <* eof
 
--- | Parser for full 'TOML' under a certain key
-subTableContent :: Key -> Parser TOML
-subTableContent key = do
+-- | Parser for a table under a certain key
+tableP :: Maybe Key -> Parser TOML
+tableP key = do
     (val, inline)  <- distributeEithers <$> many hasKeyP
     (table, array) <- fmap distributeEithers
         $ many
-        $ childKeyP key
-        $ eitherPairP (try tableP) tableArrayP
+        $ eitherPairP
+            (try $ do
+              (kDiff, k) <- childKeyP key tableNameP
+              t <- tableP (Just k)
+              pure (kDiff, t))
+            (do
+              (kDiff, k) <- childKeyP key tableArrayNameP
+              a <- tableArrayP k
+              pure (kDiff, a))
 
     pure TOML
         { tomlPairs       = HashMap.fromList val
         , tomlTables      = fromList $ inline ++ table
         , tomlTableArrays = HashMap.fromList array
         }
 
--- | @childKeyP key p@ returns the result of @p@ if the key returned by @p@ is
--- a child key of the @key@, and fails otherwise.
-childKeyP :: Key -> Parser (Key, a) -> Parser (Key, a)
-childKeyP key p = try $ do
-    (k, x) <- p
+-- | @childKeyP (Just key) p@ checks if the result of @p@ if a child key of
+-- @key@ and returns the difference of the keys and the child key.
+-- @childKeyP Nothing p@ is only called from @tomlP@ (no parent key).
+childKeyP :: Maybe Key -> Parser Key -> Parser (Key, Key)
+childKeyP Nothing parser = try $ do
+  k <- parser
+  pure (k, k)
```

Comment:
I would rewrite it as `try $ (\x -> (x, x) <$> parser`. I see why you did it this way: you prefer symmetry in pattern-matching. In that case, you can leave the implementation as it is, but you should use 4 spaces according to our style guide.

Reply:
I don't mind your suggestion, I didn't particularly like my way of writing it. And I'll re-check the spacing everywhere.